File: src/ai-services/signal-engine/app/quantitative/gold_features.py
"""
XAUUSD Feature Engineering Pipeline

Extracts institutional-grade features from raw market data for the signal model.
All features are Gold-specific — correlations, seasonality, and volatility
characteristics tuned to XAU/USD behavior.

Design principle: features must be explainable (no black-box inputs),
fast to compute (<10ms per symbol per cycle), and numerically stable
under all market conditions including extreme moves.
"""
from __future__ import annotations

import numpy as np
import pandas as pd
from dataclasses import dataclass, field
from typing import Optional


PIP = 0.01           # XAUUSD 1 pip = $0.01
MIN_TARGET = 15.0    # $15 minimum move target
MAX_TARGET = 30.0    # $30 maximum move target


@dataclass
class GoldFeatures:
    """
    Feature vector fed into both the quantitative scoring model
    and the confidence ensemble. All values are normalised to [-1, 1]
    or [0, 1] unless noted.
    """
    # --- Structure features ---
    htf_structure_score: float = 0.0      # [-100, 100] → normalised
    ltf_structure_score: float = 0.0
    htf_bullish: float = 0.0              # 1 / -1 / 0
    bos_present: float = 0.0
    choch_present: float = 0.0
    unmitigated_ob_present: float = 0.0
    ob_proximity_score: float = 0.0       # how close price is to an OB [0-1]
    fvg_present: float = 0.0
    fvg_proximity_score: float = 0.0

    # --- Liquidity features ---
    liquidity_sweep_recent: float = 0.0   # 1 if sweep in last 3 candles
    bsl_distance_pips: float = 0.0        # pips to nearest buy-side liquidity
    ssl_distance_pips: float = 0.0        # pips to nearest sell-side liquidity
    liquidity_imbalance: float = 0.0      # BSL vs SSL asymmetry

    # --- Correlation features (Gold-specific) ---
    dxy_momentum: float = 0.0            # negative → bullish Gold
    yield_momentum: float = 0.0          # negative → bullish Gold (real yields)
    vix_level: float = 0.0               # normalised VIX [0-1]
    risk_off_score: float = 0.0          # composite risk-off indicator
    gold_correlation_score: float = 0.0  # weighted macro alignment for Gold

    # --- Volatility features ---
    atr_ratio: float = 0.0               # current ATR / 20-period avg ATR
    adr_pct: float = 0.0                 # average daily range as % of price
    volatility_regime: float = 0.0       # 0=low, 0.5=normal, 1=high
    range_position: float = 0.0          # where in today's range (0=low, 1=high)

    # --- Session features ---
    session_london: float = 0.0
    session_ny: float = 0.0
    session_overlap: float = 0.0         # highest weight — peak Gold liquidity
    session_dead: float = 0.0            # penalise off-hours

    # --- News / sentiment features ---
    news_sentiment_score: float = 0.0    # [-1, 1]
    news_impact_score: float = 0.0       # [0, 1]
    high_impact_event_imminent: float = 0.0  # 1 if event < 30 min away
    event_surprise_score: float = 0.0    # actual vs forecast deviation

    # --- Momentum features ---
    momentum_1h: float = 0.0             # price change last 1H normalised by ATR
    momentum_4h: float = 0.0
    rsi_h1: float = 0.0                  # normalised RSI [0-1]
    macd_signal: float = 0.0             # MACD histogram sign and magnitude

    def to_vector(self) -> np.ndarray:
        return np.array([
            self.htf_structure_score / 100,
            self.ltf_structure_score / 100,
            self.htf_bullish,
            self.bos_present,
            self.choch_present,
            self.unmitigated_ob_present,
            self.ob_proximity_score,
            self.fvg_present,
            self.fvg_proximity_score,
            self.liquidity_sweep_recent,
            min(self.bsl_distance_pips / 5000, 1.0),
            min(self.ssl_distance_pips / 5000, 1.0),
            self.liquidity_imbalance,
            np.clip(self.dxy_momentum, -1, 1),
            np.clip(self.yield_momentum, -1, 1),
            min(self.vix_level / 50, 1.0),
            self.risk_off_score,
            self.gold_correlation_score,
            np.clip(self.atr_ratio, 0, 3) / 3,
            min(self.adr_pct / 2, 1.0),
            self.volatility_regime,
            self.range_position,
            self.session_london,
            self.session_ny,
            self.session_overlap,
            self.session_dead,
            np.clip(self.news_sentiment_score, -1, 1),
            self.news_impact_score,
            self.high_impact_event_imminent,
            np.clip(self.event_surprise_score, -2, 2) / 2,
            np.clip(self.momentum_1h, -3, 3) / 3,
            np.clip(self.momentum_4h, -3, 3) / 3,
            self.rsi_h1,
            np.clip(self.macd_signal, -1, 1),
        ], dtype=np.float32)


def extract_features(payload: dict) -> GoldFeatures:
    """Build GoldFeatures from the ASP.NET inference request payload."""
    htf = payload.get("htf_structure", {})
    ltf = payload.get("ltf_structure", {})
    corr = payload.get("correlations", {})
    vol = payload.get("volatility", {})
    news_list = payload.get("recent_news", [])
    events = payload.get("upcoming_events", [])
    session = payload.get("session", "")
    price = htf.get("current_price", 0.0)

    f = GoldFeatures()

    # ── Structure ──────────────────────────────────────────────────────────────
    f.htf_structure_score = float(htf.get("structure_score", 0))
    f.ltf_structure_score = float(ltf.get("structure_score", 0))
    f.htf_bullish = 1.0 if htf.get("bullish") else -1.0
    f.bos_present = 1.0 if htf.get("bos") else 0.0
    f.choch_present = 1.0 if htf.get("choch") else 0.0

    obs = htf.get("order_blocks", [])
    unmitigated = [ob for ob in obs if ob.get("unmitigated")]
    f.unmitigated_ob_present = 1.0 if unmitigated else 0.0

    if unmitigated and price > 0:
        nearest_ob = min(unmitigated, key=lambda ob: abs((ob["high"] + ob["low"]) / 2 - price))
        ob_mid = (nearest_ob["high"] + nearest_ob["low"]) / 2
        dist_pips = abs(price - ob_mid) / PIP
        # proximity: 1.0 = price is inside OB, 0 = >500 pips away
        f.ob_proximity_score = max(0.0, 1.0 - dist_pips / 500)

    fvgs = htf.get("fvgs", [])
    open_fvgs = [g for g in fvgs if not g.get("filled")]
    f.fvg_present = 1.0 if open_fvgs else 0.0
    if open_fvgs and price > 0:
        nearest_fvg = min(open_fvgs, key=lambda g: abs((g["upper"] + g["lower"]) / 2 - price))
        fvg_mid = (nearest_fvg["upper"] + nearest_fvg["lower"]) / 2
        dist_pips = abs(price - fvg_mid) / PIP
        f.fvg_proximity_score = max(0.0, 1.0 - dist_pips / 300)

    # ── Correlation (Gold-specific signs) ─────────────────────────────────────
    # DXY inverse: rising DXY → bearish Gold → negative feature for BUY
    dxy_chg = float(corr.get("dxy_change_1h", 0))
    f.dxy_momentum = -np.tanh(dxy_chg * 5)   # invert: DXY up → negative for Gold BUY

    # Real yields inverse: rising yields → bearish Gold
    yield_chg = float(corr.get("us10y_change_1h", 0))
    f.yield_momentum = -np.tanh(yield_chg * 10)

    vix = float(corr.get("vix", 20))
    f.vix_level = vix

    # Risk-off: high VIX + falling equities → Gold bullish
    risk_off = 1.0 if (vix > 25 or float(corr.get("spx_change_1d", 0)) < -1.0) else 0.0
    f.risk_off_score = risk_off

    # Composite Gold correlation score: sum of aligned factors
    gold_bull_factors = [
        1.0 if dxy_chg < -0.1 else 0.0,
        1.0 if yield_chg < -0.02 else 0.0,
        1.0 if risk_off else 0.0,
    ]
    f.gold_correlation_score = sum(gold_bull_factors) / 3.0

    # ── Volatility ─────────────────────────────────────────────────────────────
    atr_1h = float(vol.get("atr_1h", 8.0))
    atr_4h = float(vol.get("atr_4h", 15.0))
    adr_pct = float(vol.get("adr_pct", 0.5))
    expanding = bool(vol.get("expanding", False))

    # ATR ratio: current ATR vs long-run baseline ($10 for H1 Gold)
    f.atr_ratio = atr_1h / 10.0
    f.adr_pct = adr_pct
    f.volatility_regime = (
        0.0 if atr_1h < 5.0 else
        1.0 if atr_1h > 20.0 else
        (atr_1h - 5.0) / 15.0
    )

    # ── Session ────────────────────────────────────────────────────────────────
    utc_hour = __import__("datetime").datetime.utcnow().hour
    f.session_london = 1.0 if 7 <= utc_hour <= 12 else 0.0
    f.session_ny = 1.0 if 13 <= utc_hour <= 20 else 0.0
    f.session_overlap = 1.0 if 13 <= utc_hour <= 16 else 0.0
    f.session_dead = 1.0 if (utc_hour >= 22 or utc_hour <= 2) else 0.0

    # ── News ───────────────────────────────────────────────────────────────────
    if news_list:
        sentiments = [n.get("sentiment", 0.0) for n in news_list]
        impacts = [n.get("impact_score", 0.0) for n in news_list]
        f.news_sentiment_score = float(np.average(sentiments, weights=[abs(s) + 0.1 for s in sentiments]))
        f.news_impact_score = float(max(impacts)) if impacts else 0.0

    imminent = [e for e in events if e.get("minutes_until", 999) <= 30 and e.get("impact") in ("High", "Critical")]
    f.high_impact_event_imminent = 1.0 if imminent else 0.0

    return f

File: src/frontend/lib/signal-engine/gold-features.ts
import type { CandleData } from '@/lib/market-data'
import type { MarketStructure } from './market-structure'

export interface GoldFeatures {
  htfStructureScore: number
  ltfStructureScore: number
  htfBullish: number           // 1 / -1
  bosPresent: number
  chochPresent: number
  unmitigatedObPresent: number
  obProximityScore: number     // [0-1]
  fvgPresent: number
  fvgProximityScore: number
  liquiditySweepRecent: number
  bslDistancePips: number
  sslDistancePips: number
  liquidityImbalance: number
  dxyMomentum: number
  yieldMomentum: number
  vixLevel: number
  riskOffScore: number
  riskOnScore: number          // inverse of riskOff: low VIX + rising SPX = bearish gold
  goldCorrelationScore: number
  atrRatio: number
  adrPct: number
  volatilityRegime: number
  rangePosition: number
  sessionLondon: number
  sessionNy: number
  sessionOverlap: number
  sessionDead: number
  newsSentimentScore: number
  newsImpactScore: number
  highImpactEventImminent: number
  eventSurpriseScore: number
  momentum1h: number
  momentum4h: number
  rsiH1: number
  macdSignal: number
}

export function extractFeatures(
  htf: MarketStructure,
  ltf: MarketStructure,
  htfScore: number,
  ltfScore: number,
  correlations: {
    dxyChange1H: number
    us10YChange1H: number
    vix: number
    spxChange1D: number
    isRiskOff: boolean
  },
  volatility: {
    atr1H: number
    atrPercent?: number
    adrPercent: number
    isExpanding: boolean
  },
  candles: CandleData[],
): GoldFeatures {
  const price = htf.currentPrice || (candles.length ? candles[candles.length - 1].close : 0)
  const atr   = volatility.atr1H

  const f: GoldFeatures = {
    htfStructureScore: htfScore,
    ltfStructureScore: ltfScore,
    htfBullish: htf.bullishStructure ? 1 : -1,
    bosPresent:  htf.breakOfStructure ? 1 : 0,
    chochPresent: htf.changeOfCharacter ? 1 : 0,
    unmitigatedObPresent: 0, obProximityScore: 0,
    fvgPresent: 0, fvgProximityScore: 0,
    liquiditySweepRecent: 0, bslDistancePips: 0, sslDistancePips: 0, liquidityImbalance: 0,
    dxyMomentum: 0, yieldMomentum: 0, vixLevel: 20,
    riskOffScore: 0, riskOnScore: 0, goldCorrelationScore: 0.33,
    atrRatio: 1, adrPct: 0.5, volatilityRegime: 0.5, rangePosition: 0.5,
    sessionLondon: 0, sessionNy: 0, sessionOverlap: 0, sessionDead: 0,
    newsSentimentScore: 0, newsImpactScore: 0, highImpactEventImminent: 0, eventSurpriseScore: 0,
    momentum1h: 0, momentum4h: 0, rsiH1: 0.5, macdSignal: 0,
  }

  // ── Order blocks (ATR-adaptive proximity) ───────────────────────────────────
  const unmitigated = htf.orderBlocks.filter(ob => ob.isUnmitigated)
  f.unmitigatedObPresent = unmitigated.length > 0 ? 1 : 0
  if (unmitigated.length && price > 0) {
    const nearest = unmitigated.reduce((a, b) =>
      Math.abs((a.high + a.low) / 2 - price) < Math.abs((b.high + b.low) / 2 - price) ? a : b)
    const distDollar = Math.abs(price - (nearest.high + nearest.low) / 2)
    // Max proximity window = 1.5× ATR (e.g. $15 at ATR=$10); floor at $5 for low-ATR conditions
    const maxDist = Math.max(5, atr * 1.5)
    f.obProximityScore = Math.max(0, 1 - distDollar / maxDist)
  }

  // ── FVGs (ATR-adaptive proximity) ────────────────────────────────────────────
  const openFvgs = htf.fairValueGaps.filter(g => !g.isFilled)
  f.fvgPresent = openFvgs.length > 0 ? 1 : 0
  if (openFvgs.length && price > 0) {
    const nearest = openFvgs.reduce((a, b) =>
      Math.abs((a.upperBound + a.lowerBound) / 2 - price) <
      Math.abs((b.upperBound + b.lowerBound) / 2 - price) ? a : b)
    const distDollar = Math.abs(price - (nearest.upperBound + nearest.lowerBound) / 2)
    const maxDist = Math.max(4, atr * 1.2)
    f.fvgProximityScore = Math.max(0, 1 - distDollar / maxDist)
  }

  // ── Liquidity ───────────────────────────────────────────────────────────────
  f.liquiditySweepRecent = htf.liquidityLevels.some(l => l.isSwept) ? 1 : 0
  const bslLevels = htf.liquidityLevels.filter(l => l.description.includes('BSL'))
  const sslLevels = htf.liquidityLevels.filter(l => l.description.includes('SSL'))
  const PIP = 0.01
  if (bslLevels.length)
    f.bslDistancePips = Math.min(...bslLevels.map(l => Math.abs(l.price - price) / PIP))
  if (sslLevels.length)
    f.sslDistancePips = Math.min(...sslLevels.map(l => Math.abs(l.price - price) / PIP))
  const total = bslLevels.length + sslLevels.length
  f.liquidityImbalance = total > 0 ? (bslLevels.length - sslLevels.length) / total : 0

  // ── Macro (Gold-specific inverse signs) ─────────────────────────────────────
  // DXY up → Gold down → negative for BUY
  f.dxyMomentum   = -Math.tanh(correlations.dxyChange1H * 5)
  // Yields up → Gold down → negative for BUY
  f.yieldMomentum = -Math.tanh(correlations.us10YChange1H * 10)
  f.vixLevel      = correlations.vix

  // Risk-off: VIX spike OR SPX sharp drop → safe-haven demand → gold bullish
  f.riskOffScore = (correlations.vix > 25 || correlations.spxChange1D < -1.0) ? 1 : 0

  // Risk-on: VIX calm AND SPX rising → risk appetite → gold bearish
  // Score grades from 0 to 1 based on how "risk-on" the environment is
  const vixContrib = correlations.vix > 0 && correlations.vix < 18
    ? Math.max(0, (18 - correlations.vix) / 18)
    : 0
  const spxContrib = correlations.spxChange1D > 0.5
    ? Math.min(1, correlations.spxChange1D / 2)
    : 0
  f.riskOnScore = Math.min(1, (vixContrib + spxContrib) / 2)

  const dxyBull   = correlations.dxyChange1H < -0.1 ? 1 : 0
  const yieldBull = correlations.us10YChange1H < -0.02 ? 1 : 0
  f.goldCorrelationScore = (dxyBull + yieldBull + (f.riskOffScore > 0 ? 1 : 0)) / 3

  // ── Volatility (calibrated for gold ATR range $5–35) ─────────────────────────
  f.atrRatio       = atr / 12   // normalised around $12 typical gold H1 ATR
  f.adrPct         = volatility.adrPercent
  // Linear mapping: ATR<5 → 0 (dead), ATR>30 → 1 (extreme); normal ~$10-15 → 0.2-0.5
  f.volatilityRegime = atr < 5 ? 0 : atr > 30 ? 1 : (atr - 5) / 25

  // ── Session ─────────────────────────────────────────────────────────────────
  const h = new Date().getUTCHours()
  f.sessionLondon  = h >= 7 && h <= 12 ? 1 : 0
  f.sessionNy      = h >= 13 && h <= 20 ? 1 : 0
  f.sessionOverlap = h >= 13 && h <= 16 ? 1 : 0
  f.sessionDead    = h >= 22 || h <= 2 ? 1 : 0

  // ── Candle-based momentum ───────────────────────────────────────────────────
  if (candles.length >= 5) {
    const last  = candles[candles.length - 1]
    const prev1 = candles[candles.length - 2]
    const prev4 = candles[Math.max(0, candles.length - 5)]
    f.momentum1h = (last.close - prev1.close) / Math.max(atr, 0.01)
    f.momentum4h = (last.close - prev4.close) / Math.max(atr, 0.01)

    const gains: number[] = [], losses: number[] = []
    const rsiCandles = candles.slice(-15)
    for (let i = 1; i < rsiCandles.length; i++) {
      const d = rsiCandles[i].close - rsiCandles[i - 1].close
      if (d > 0) gains.push(d); else losses.push(-d)
    }
    const avgGain = gains.reduce((s, v) => s + v, 0) / 14
    const avgLoss = losses.reduce((s, v) => s + v, 0) / 14
    const rs = avgLoss === 0 ? 100 : avgGain / avgLoss
    f.rsiH1 = (100 - 100 / (1 + rs)) / 100
  }

  return f
}
